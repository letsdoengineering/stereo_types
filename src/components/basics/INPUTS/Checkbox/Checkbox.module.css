/* Customize the label (the container) */
.container {
    display: flex;
    align-items: center;
    position: relative;
    margin-left: 16px;
    cursor: pointer;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
}

/* Hide the browser's default checkbox */
.container input {
    position: absolute;
    opacity: 0;
    cursor: pointer;
    height: 0;
    width: 0;
}

/* Add style for when focus on original input style to our checkmark */
.container input:focus ~ .checkmark {
    -moz-box-shadow:    0 0 0 2px var(--focus);
    -webkit-box-shadow: 0 0 0 2px var(--focus);
    box-shadow:         0 0 0 2px var(--focus);
}

.label {
    font-size: var(--font-size-S);
    color: var(--dark-gray);
}
/* Create a custom checkbox */
.checkmark {
    margin-right: 4px;
    height: 1rem;
    width: 1rem;
    border: 2px solid var(--dark-gray);
    background-color: white;
}

/* When checked add background */
.container input:checked ~ .checkmark {
    background-color: var(--dark-gray);
}

/* When checked but unavailable add gray background */
.container input:checked ~ .checkmark.unavailable {
    background-color: var(--outline-gray);
}

/* Create the checkmark/indicator (hidden when not checked) */
.checkmark:after {
    content: "";
    position: absolute;
    display: none;
}

/* Show the checkmark when checked */
.container input:checked ~ .checkmark:after {
    display: block;
}

/* Style the checkmark/indicator */
.container .checkmark:after {
    left: 5px;
    width: 6px;
    height: 10px;
    border: solid white;
    border-width: 0 3px 3px 0;
    -webkit-transform: rotate(45deg);
    -ms-transform: rotate(45deg);
    transform: rotate(45deg);
}

.unavailable {
    color: var(--outline-gray);
    border-color: var(--outline-gray);
    cursor: default;
}
