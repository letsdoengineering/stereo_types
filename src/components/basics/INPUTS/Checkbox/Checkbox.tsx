import React, { FunctionComponent } from 'react'
import Text from '../../Text/Text'
import * as checkboxStyles from './Checkbox.module.css'
import getClassNames from '../../../../utils/get-class-names'

export type CheckboxProps = React.ComponentProps<'input'> &
    Readonly<{
        /** id used for id of input - custom styling */
        id?: string | undefined
        /** tell the checkbox what it's status is, controlled by consumer and triggered via onChange */
        isChecked: boolean
        /** Called when toggled */
        onChange(): void
        /** value you wish the checkbox to have */
        value: string
        /** text for showing next to checkbox */
        text: string
        /** toggle to make unavailable */
        available?: boolean
    }>

const Checkbox: FunctionComponent<CheckboxProps> = ({
    isChecked,
    id,
    available = true,
    onChange,
    value,
    text,
    ...rest
}) => {
    const textClassName = getClassNames({
        conditionalClasses: { [`${checkboxStyles.unavailable}`]: !available },
    })
    const checkboxClassName = getClassNames({
        defaultClasses: [`${checkboxStyles.checkmark}`],
        conditionalClasses: { [`${checkboxStyles.unavailable}`]: !available },
    })

    return (
        <label className={checkboxStyles.container}>
            <input
                disabled={!available}
                checked={isChecked}
                id={id}
                onChange={onChange}
                type='checkbox'
                value={value}
                {...rest}
            />
            <span className={checkboxClassName} />
            <Text className={textClassName}>{text}</Text>
        </label>
    )
}

export default Checkbox
