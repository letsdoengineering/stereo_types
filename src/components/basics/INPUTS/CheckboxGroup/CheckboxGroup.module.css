.container {
    border: none;
    padding: 0;
    margin: 4px;
}

.container label {
    margin-top: 3px;
}
