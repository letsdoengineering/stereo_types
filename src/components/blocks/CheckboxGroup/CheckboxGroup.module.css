.container {
    border: none;
    padding: 0;
    margin: 0;
}

.container label {
    margin-top: 0.25rem;
}
