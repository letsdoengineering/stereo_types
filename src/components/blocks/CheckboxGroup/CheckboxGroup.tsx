import React from 'react'

import getClassNames from '../../../utils/get-class-names'
import Text from '../../basics/Text/Text'
import * as checkboxGroupStyles from './CheckboxGroup.module.css'

type CheckboxGroupProps = Readonly<{
    /** label must be provided for accessibility */
    groupLabel: string
    /** label is visually hidden by default */
    showLabel?: boolean
    /** children always of type ReactNode */
    children: React.ReactNode
}>

const CheckboxGroup: React.FC<CheckboxGroupProps> = ({
    groupLabel,
    showLabel = false,
    children,
}) => {
    const groupLabelClassName = getClassNames({
        conditionalClasses: {
            ['visually-hidden']: !showLabel,
        },
    })

    return (
        <form>
            <fieldset className={checkboxGroupStyles.container}>
                <legend className={groupLabelClassName}>
                    <Text>{groupLabel}</Text>
                </legend>
                {children}
            </fieldset>
        </form>
    )
}

export default CheckboxGroup
