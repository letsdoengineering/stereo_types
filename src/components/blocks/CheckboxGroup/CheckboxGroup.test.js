import React from 'react'
import { render } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { axe } from 'jest-axe'
import Checkbox from '../../basics/INPUTS/Checkbox/Checkbox'
import CheckboxGroup from './CheckboxGroup'

describe('CheckboxGroup', () => {
    const mockOnChange = jest.fn()
    const options = ['Cake', 'Cheese', 'Potatoes']

    const checkboxes = options.map((option, i) => (
        <Checkbox
            key={`option-${option}-${i}`}
            value={option}
            isChecked={true}
            onChange={() => mockOnChange()}
        />
    ))

    it('should render as default expected', () => {
        const tree = renderer
            .create(
                <CheckboxGroup groupLabel={'A group of checkboxes'}>{checkboxes}</CheckboxGroup>
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('should render as expected with showLabel=true', () => {
        const tree = renderer
            .create(
                <CheckboxGroup groupLabel={'A group of checkboxes'} showLabel>
                    {checkboxes}
                </CheckboxGroup>
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders the page DOM that is accessible', async () => {
        const { container } = render(
            <CheckboxGroup groupLabel={'A group of checkboxes'}>{checkboxes}</CheckboxGroup>
        )
        const results = await axe(container)

        expect(results).toHaveNoViolations()
    })
})
