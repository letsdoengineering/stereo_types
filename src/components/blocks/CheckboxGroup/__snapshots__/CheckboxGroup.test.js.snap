// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup should render as default expected 1`] = `
<form>
  <fieldset
    className="container"
  >
    <legend
      className="visually-hidden"
    >
      <span
        className="sizeS weightNormal textGray"
      >
        A group of checkboxes
      </span>
    </legend>
    <label
      className="container"
    >
      <input
        checked={true}
        disabled={false}
        onChange={[Function]}
        type="checkbox"
        value="Cake"
      />
      <span
        className="checkmark"
      />
      <span
        className="sizeS weightNormal textGray"
      >
        Cake
      </span>
    </label>
    <label
      className="container"
    >
      <input
        checked={true}
        disabled={false}
        onChange={[Function]}
        type="checkbox"
        value="Cheese"
      />
      <span
        className="checkmark"
      />
      <span
        className="sizeS weightNormal textGray"
      >
        Cheese
      </span>
    </label>
    <label
      className="container"
    >
      <input
        checked={true}
        disabled={false}
        onChange={[Function]}
        type="checkbox"
        value="Potatoes"
      />
      <span
        className="checkmark"
      />
      <span
        className="sizeS weightNormal textGray"
      >
        Potatoes
      </span>
    </label>
  </fieldset>
</form>
`;

exports[`CheckboxGroup should render as expected with showLabel=true 1`] = `
<form>
  <fieldset
    className="container"
  >
    <legend
      className=""
    >
      <span
        className="sizeS weightNormal textGray"
      >
        A group of checkboxes
      </span>
    </legend>
    <label
      className="container"
    >
      <input
        checked={true}
        disabled={false}
        onChange={[Function]}
        type="checkbox"
        value="Cake"
      />
      <span
        className="checkmark"
      />
      <span
        className="sizeS weightNormal textGray"
      >
        Cake
      </span>
    </label>
    <label
      className="container"
    >
      <input
        checked={true}
        disabled={false}
        onChange={[Function]}
        type="checkbox"
        value="Cheese"
      />
      <span
        className="checkmark"
      />
      <span
        className="sizeS weightNormal textGray"
      >
        Cheese
      </span>
    </label>
    <label
      className="container"
    >
      <input
        checked={true}
        disabled={false}
        onChange={[Function]}
        type="checkbox"
        value="Potatoes"
      />
      <span
        className="checkmark"
      />
      <span
        className="sizeS weightNormal textGray"
      >
        Potatoes
      </span>
    </label>
  </fieldset>
</form>
`;
